sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -367,18 +371,47 @@ pi_result cuda_piEventRetain(pi_event event);
 
 /// \endcond
 
-CUstream _pi_queue::get_next_compute_stream() {
-  if (num_compute_streams_ < compute_streams_.size()) {
-    // the check above is for performance - so as not to lock mutex every time
-    std::lock_guard<std::mutex> guard(compute_stream_mutex_);
-    // The second check is done after mutex is locked so other threads can not
-    // change num_compute_streams_ after that
+CUstream _pi_queue::get_next_compute_stream(pi_uint32 *stream_token) {
+  pi_uint32 stream_i;
+  while (true) {
     if (num_compute_streams_ < compute_streams_.size()) {
-      PI_CHECK_ERROR(
-          cuStreamCreate(&compute_streams_[num_compute_streams_++], flags_));
+      // the check above is for performance - so as not to lock mutex every time
+      std::lock_guard<std::mutex> guard(compute_stream_mutex_);
+      // The second check is done after mutex is locked so other threads can not
+      // change num_compute_streams_ after that
+      if (num_compute_streams_ < compute_streams_.size()) {
+        PI_CHECK_ERROR(
+            cuStreamCreate(&compute_streams_[num_compute_streams_++], flags_));
+      }
+    }
+    stream_i = compute_stream_idx_++;
+    if (delay_compute_[stream_i % compute_streams_.size()]) {
+      delay_compute_[stream_i % compute_streams_.size()] = 0;
+    } else {
+      break;
     }
   }
-  return compute_streams_[compute_stream_idx_++ % compute_streams_.size()];
+  if (stream_token) {
+    *stream_token = stream_i;
+  }
+  return compute_streams_[stream_i % compute_streams_.size()];
+}
+
+CUstream getComputeStream(pi_queue queue, pi_uint32 num_events_in_wait_list,
+                          const pi_event *event_wait_list,
+                          pi_uint32 *stream_token = nullptr) {
+  for (pi_uint32 i = 0; i < num_events_in_wait_list; i++) {
+    pi_uint32 token = event_wait_list[i]->get_stream_token();
+    if (event_wait_list[i]->get_queue() == queue &&
+        queue->is_last_command(token)) {
+      queue->delay_stream(token);
+      if (stream_token) {
+        *stream_token = token;
+      }
+      return event_wait_list[i]->get_stream();
```

Comment:
Doesn't getting streams out-of-order interfere with the optimizations in https://github.com/intel/llvm/pull/6201? Depending on which is merged first, should `sync_streams` go through missed delayed streams and sync with those too?

Reply:
You are right, this will need to be redesigned.

Reply:
https://github.com/intel/llvm/pull/6201 has been merged, so I think here would be the place to do it. 😄

Reply:
This should be fixed now.

Reply:
It looks that way! As an additional note, shouldn't `sync_streams` also clear `delay_compute_` for the streams it syncs?